text/0000-std-function-macro.md

```diff
@@ -0,0 +1,153 @@
+- Feature Name: Function Name Macro
+- Start Date: (2016-08-19)
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+`function_name` is a macro, which expands to the fully qualified function name
+from which it was invoked. It works similar to `std::file`, `std::line` or
+`std::module_path` and expands to an expression of type `&'static str`.
+
+# Motivation
+[motivation]: #motivation
+
+There are currently macros for getting the source file name, `std::file`, and
+the line number, `std::line`. To complete this pair, a macro to for getting
+the current function name would be useful. All three macros are useful for
+e.g. generating log output with more context about where something was
+happening. The file and line is good for being able to look up the exact
+location, however the fully qualified function name gives additional context
+without even having to look into the source code.
+
+# Detailed design
+[design]: #detailed-design
+
+This feature would have an equivalent design and implementation to
+`std::file`. It would return a fully qualified function name but the actual
+string returned is implementation defined. The string is meant to be used by
+the user to get an idea about the context of the code in e.g. debug output,
+and should be meaningful for a human, but is not meant to be analyzed by a
+program.
+
+For example the following scheme could be used:
+
+ - Plain functions use their name, e.g. `some_function` for
+
+````
+    fn some_function(arg: SomeThing) {
+    ...
+    }
+````
+
+ - Generic functions use their name, followed by the generic arguments, e.g.
+   `some_function<T, U>` for
```

Comment:
> These kind of thoughts are exactly why the exact implementation is considered implementation defined, 

This is hardly something we can afford doing for a stable macro. Its output will become de fact stable no matter what documentation says.

> Do you have an example where avoiding monomorphisation wouldn't allow uniquely identifying such a function?

Something that needs monomorphisation/full set of bounds to differentiate:

```
#![feature(specialization)]

trait A {
    fn a(&self) {}
}

impl<I> A for I {
    default fn a(&self) { 
        println!("I::a()") // {Concrete type}::a()? <_ as I: !Display>::a()?
    }
}

impl<I: ::std::fmt::Display> A for I {
    fn a(&self) { 
        println!("<_ as I: Display>::a()") // {Concrete type}::a()?
    }
}

fn main() {
    <i32 as A>::a(&0);
    <() as A>::a(&());
}
```

---

Something that cannot be differentiated without UFCS:

```
trait A { fn a(&self) {} }
trait B { fn a() -> Self where Self: Sized; }

struct C;

impl A for C {}
impl B for C { fn a() -> C { C } }

fn main() {
    <C as A>::a(&<C as B>::a());
}
```

So in general I think that the attempt itself to uniquely identify functions by a path is not a very feasible business if you also want to always receive a nice path.